internal/backends/nodejs/nodejs.go

```diff
@@ -48,24 +50,124 @@ type npmSearchResults struct {
+//
+// which means that we have to deserialize the union of the type described in the document, and the
+// type of a package.json file: https://docs.npmjs.com/cli/v10/configuring-npm/package-json
 type npmInfoResult struct {
-	Name     string                 `json:"name"`
-	Versions map[string]interface{} `json:"versions"`
-	Author   struct {
-		Name  string `json:"name"`
-		Email string `json:"email"`
-		URL   string `json:"url"`
-	} `json:"author"`
-	Bugs struct {
-		URL string `json:"url"`
-	} `json:"bugs"`
-	Description string `json:"description"`
-	Homepage    string `json:"homepage"`
-	License     string `json:"license"`
-	Repository  struct {
-		Type string `json:"type"`
-		URL  string `json:"url"`
-	} `json:"repository"`
+	Name        string                 `json:"name"`
+	Versions    map[string]interface{} `json:"versions"`
+	Author      packageJsonPerson      `json:"author"`
+	Bugs        packageJsonBugs        `json:"bugs"`
+	Description string                 `json:"description"`
+	Homepage    string                 `json:"homepage"`
+	License     string                 `json:"license"`
+	Repository  packageJsonRepository  `json:"repository"`
+}
+
+type packageJsonRepository struct {
+	URL string
+}
+
+func (repo *packageJsonRepository) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		repo.URL = str
+		return nil
+	}
+
+	var obj map[string]interface{}
+	if err := json.Unmarshal(data, &obj); err == nil {
+		if url, ok := obj["url"].(string); ok {
+			repo.URL = url
+		}
+		return nil
+	}
+
+	return fmt.Errorf("expected repository in package.json to be a string or an object")
+}
+
+type packageJsonPerson struct {
+	Name  string
+	Email string
+	URL   string
+}
+
+var packageJsonPersonStringRegexp *regexp.Regexp
+
+func (person *packageJsonPerson) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		if packageJsonPersonStringRegexp == nil {
+			// parses the form: name <email> (url)
+			// eg: John Doe <joe.dohn@example.zip> (ftp://C:\Nix\Store)
+			// any of the 3 are optional
+			packageJsonPersonStringRegexp = regexp.MustCompile(`([^\<\(]*)?\s*(\<.*\>)?\s*(\(.*\))?`)
+		}
+
+		captures := packageJsonPersonStringRegexp.FindStringSubmatch(str)
+		if len(captures) == 4 {
+			if name := captures[1]; name != "" {
+				person.Name = name
+			}
+
+			if email := captures[2]; email != "" {
+				person.Email = email
+			}
+
+			if url := captures[3]; url != "" {
+				person.URL = url
+			}
+
+			return nil
+		}
+	}
+
+	var obj map[string]interface{}
+	if err := json.Unmarshal(data, &obj); err == nil {
+		if name, ok := obj["name"].(string); ok {
+			person.Name = name
+		}
+
+		if email, ok := obj["email"].(string); ok {
+			person.Email = email
+		}
+
+		if url, ok := obj["url"].(string); ok {
+			person.URL = url
+		}
+
+		return nil
+	}
+
+	return errors.New("expected person in package.json to be a string or an object")
```

Comment:
If we fail to parse the author, can we continue with an empty author instead of failing? I'm worried about the string regex failing to parse some author strings and then we still fail.

Reply:
i can add a fallback in the path where the regexp fails to parse! i don't feel comfortable changing this return though, as it's the case where the author field is neither a string nor an object, which _should_ be an error. handling deserialization errors is the responsibility further up the stack from this function, we shouldn't expect this function to always succeed

Reply:
Sounds good.